src/status_im/ui/screens/wallet/send/events.cljs

```diff
@@ -385,3 +391,109 @@
+                              (:message error)))
+          (reset! in-progress? false))
+      (do
+        (re-frame/dispatch [:wallet/add-unconfirmed-transaction transaction result])
+        (if on-result
+          (re-frame/dispatch (conj on-result id result method))
+          (re-frame/dispatch [:send-transaction-message public-key {:address to
+                                                                    :asset   (name symbol)
+                                                                    :amount  amount-text
+                                                                    :tx-hash result}]))))))
+
+(defn send-transaction-wrapper [{:keys [transaction password flow all-tokens in-progress? chain contact account]}]
+  (let [symbol (:symbol transaction)
+        coin   (tokens/asset-for all-tokens (keyword chain) symbol)]
+    (reset! in-progress? true)
+    (send-transaction! (models.wallet/prepare-send-transaction (:address account) transaction)
+                       symbol
+                       coin
+                       #(on-transaction-completed transaction flow contact coin (types/json->clj %) in-progress?)
+                       password)))
+
+(defn qr-data->send-tx-data [{:keys [address value symbol gas gasPrice public-key from-chat?]}]
+  {:pre [(not (nil? address))]}
+  (cond-> {:to address :public-key public-key}
+    value (assoc :amount value)
+    symbol (assoc :symbol symbol)
+    gas (assoc :gas (money/bignumber gas))
+    from-chat? (assoc :from-chat? from-chat?)
+    gasPrice (assoc :gas-price (money/bignumber gasPrice))))
+
+(defn extract-qr-code-details [chain all-tokens qr-uri]
+  {:pre [(keyword? chain) (string? qr-uri)]}
+  ;; i don't like fetching all tokens here
+  (let [qr-uri (string/trim qr-uri)
+        chain-id (ethereum/chain-keyword->chain-id chain)]
+    (or (let [m (eip681/parse-uri qr-uri)]
+          (merge m (eip681/extract-request-details m all-tokens)))
+        (when (ethereum/address? qr-uri)
+          {:address qr-uri :chain-id chain-id}))))
+
+(defn qr-data->transaction-data [qr-data contacts]
+  {:pre [(map? qr-data)]}
+  (let [{:keys [to name] :as tx-details} (qr-data->send-tx-data qr-data)
+        contact-name (:name (contact.db/find-contact-by-address contacts to))]
+    (cond-> (assoc tx-details :qr true)
+      contact-name (assoc :to-name name))))
+
+;; CHOOSEN RECIPIENT
+(defn eth-name->address [chain recipient callback]
+  (if (ens/is-valid-eth-name? recipient)
+    (ens/get-addr (get ens/ens-registries chain)
+                  recipient
+                  #(callback %))
+    (callback recipient)))
+
+(defn chosen-recipient [chain {:keys [to to-ens qr] :as params} success-callback error-callback]
+  {:pre [(keyword? chain) (string? to)]}
+  (eth-name->address chain to
+                     (fn [to]
+                       (if (ethereum/address? to)
+                         (if (and (not qr) (not to-ens) (not (eip55/valid-address-checksum? to)))
```

Comment:
`(not (or qr to-ens (eip55/valid-address-checksum? to)))` would have less `not`, probably less readable though, idk

Reply:
I too thought it would've been less readable even though it's more concise. The way it is now corresponds to how you'd explain it out loud in English, so I kept it like that

Reply:
understood